app/models/user.rb

```diff
@@ -21,9 +21,30 @@ class User < ApplicationRecord
   # :confirmable, :lockable, :timeoutable and :omniauthable
   # https://github.com/plataformatec/devise
   devise :database_authenticatable, :registerable,
-         :recoverable, :rememberable, :validatable, :omniauthable, omniauth_providers: [:hpiopenid]
+         :recoverable, :rememberable, :validatable, :omniauthable, omniauth_providers: [:hpiopenid], password_length: 8..128
 
-  validates :first_name, presence: true
+  validate :password_complexity
+
+  def password_complexity
+    if password.present?
+      not_only_numbers
+      four_different_characters
+    end
+  end
+
+  def not_only_numbers
+    unless password.match(/^(?!^\d+$)^.+$/)
+      errors.add :password, I18n.t('validations.not_only_numbers')
+    end
+  end
+
+  def four_different_characters
+    if password.chars.uniq.count < 4
+      errors.add :password, I18n.t('validations.four_different_characters')
+    end
+  end
+
+  validates :first_name, :last_name, presence: true
```

Comment:
I think you usually put all method declarations below all validation declaration